geo/src/algorithm/is_convex.rs

```diff
@@ -0,0 +1,248 @@
+    }
+
+    /// Test if the shape is convex, and oriented
+    /// counter-clockwise.
+    fn is_ccw_convex(&self) -> bool {
+        self.convex_orientation(true, Some(Orientation::CounterClockwise))
+            .is_some()
+    }
+
+    /// Test if the shape is convex, and oriented clockwise.
+    fn is_cw_convex(&self) -> bool {
+        self.convex_orientation(true, Some(Orientation::Clockwise))
+            .is_some()
+    }
+
+    /// Test if the shape is strictly convex.
+    fn is_strictly_convex(&self) -> bool {
+        self.convex_orientation(false, None).is_some()
+    }
+
+    /// Test if the shape is strictly convex, and oriented
+    /// counter-clockwise.
+    fn is_strictly_ccw_convex(&self) -> bool {
+        self.convex_orientation(false, Some(Orientation::CounterClockwise))
+            == Some(Orientation::CounterClockwise)
+    }
+
+    /// Test if the shape is strictly convex, and oriented
+    /// clockwise.
+    fn is_strictly_cw_convex(&self) -> bool {
+        self.convex_orientation(false, Some(Orientation::Clockwise)) == Some(Orientation::Clockwise)
+    }
+
+    /// Test if the shape lies on a line.
+    fn is_collinear(&self) -> bool;
+}
+
+impl<T: CoordinateType + HasKernel> IsConvex for LineString<T> {
+    fn convex_orientation(
+        &self,
+        allow_collinear: bool,
+        specific_orientation: Option<Orientation>,
+    ) -> Option<Orientation> {
+        if !self.is_closed() || self.0.is_empty() {
+            None
+        } else {
+            is_convex_shaped(&self.0[1..], allow_collinear, specific_orientation)
+        }
+    }
+
+    fn is_collinear(&self) -> bool {
+        self.0.is_empty()
+            || is_convex_shaped(&self.0[1..], true, Some(Orientation::Collinear)).is_some()
+    }
+}
+
+/// A utility that tests convexity of a sequence of
+/// coordinates. It verifies that for all `0 <= i < n`, the
+/// vertices at positions `i`, `i+1`, `i+2` (mod `n`) have
+/// the same orientation, optionally accepting collinear
+/// triplets, and expecting a specific orientation. The
+/// output is `None` or the only non-collinear orientation,
+/// unless everything is collinear.
+fn is_convex_shaped<T>(
+    coords: &[Coordinate<T>],
+    allow_collinear: bool,
+    specific_orientation: Option<Orientation>,
+) -> Option<Orientation>
+where
+    T: CoordinateType + HasKernel,
+{
+    let n = coords.len();
+
+    let orientation_at = |i: usize| {
+        let coord = coords[i];
+        let next = coords[(i + 1) % n];
+        let nnext = coords[(i + 2) % n];
+        (i, T::Ker::orient2d(coord, next, nnext))
+    };
+
+    let find_first_non_collinear = (0..n).map(orientation_at).find_map(|(i, orientation)| {
+        match orientation {
+            Orientation::Collinear => {
+                // If collinear accepted, we skip, otherwise
+                // stop.
+                if allow_collinear {
+                    None
+                } else {
+                    Some((i, orientation))
```

Comment:
A little clarity nit: 

At this point,  find_first_non_collinear is in fact collinear, right?

To me that reads like `let not_a_dog = 🐕;` 

Would something like `starting_orientation` be clearer than `find_first_non_collinear`?

No biggie though, it just tripped me up as I was trying to understand.

Reply:
So, the first loop is in fact finding the first non-collinear orientation;  we return collinear in that loop only if !allow_collinear, and as a fast exit path.  The apparent inversion is unfortunately how `find` / `find_map` works:  it shorts at the first `Some()` return from the closure.